geode-core/src/main/java/org/apache/geode/internal/cache/control/HeapMemoryMonitor.java

```diff
@@ -144,82 +84,34 @@
   private final ResourceAdvisor resourceAdvisor;
   private final InternalCache cache;
   private final ResourceManagerStats stats;
-  private final TenuredHeapConsumptionMonitor tenuredHeapConsumptionMonitor;
 
-  @MutableForTesting
-  private static boolean testDisableMemoryUpdates = false;
   @MutableForTesting
   private static long testBytesUsedForThresholdSet = -1;
 
-  /**
-   * Determines if the name of the memory pool MXBean provided matches a list of known tenured pool
-   * names.
-   *
-   * Package private for testing.
-   * checkTenuredHeapConsumption
-   *
-   * @param memoryPoolMXBean The memory pool MXBean to check.
-   * @return True if the pool name matches a known tenured pool name, false otherwise.
-   */
-  static boolean isTenured(MemoryPoolMXBean memoryPoolMXBean) {
-    if (memoryPoolMXBean.getType() != MemoryType.HEAP) {
-      return false;
-    }
-
-    String name = memoryPoolMXBean.getName();
-
-    return name.equals("CMS Old Gen") // Sun Concurrent Mark Sweep GC
-        || name.equals("PS Old Gen") // Sun Parallel GC
-        || name.equals("G1 Old Gen") // Sun G1 GC
-        || name.equals("Old Space") // BEA JRockit 1.5, 1.6 GC
-        || name.equals("Tenured Gen") // Hitachi 1.5 GC
-        || name.equals("Java heap") // IBM 1.5, 1.6 GC
-        || name.equals("GenPauseless Old Gen") // azul C4/GPGC collector
-
-        // Allow an unknown pool name to monitor
-        || (HEAP_POOL != null && name.equals(HEAP_POOL));
-  }
+  @Immutable
+  private static final ServiceLoader<HeapUsageProvider> heapUsageMonitorLoader =
+      ServiceLoader.load(HeapUsageProvider.class);
```

Comment:
This approach seems quite limiting and not at all conducive dependency injection. Also it would make it really hard to test, without having to make sure that the correct jars are on the classpath.

Maybe the approach could be amended, where the constructor either has the HeapUsageProvider provided as an argument OR an identifier is passed in, which allows the constructor to find the suitable HeapUsageProvider (just in case there are more than 1 on the classpath).

Reply:
Good idea.